src/solidity-frontend/solidity_convert.cpp

```diff
@@ -0,0 +1,1898 @@
+
+    // wrap it in an ImplicitCastExpr to perform conversion of FunctionToPointerDecay
+    nlohmann::json implicit_cast_expr =
+      make_implicit_cast_expr(callee_expr_json, "FunctionToPointerDecay");
+    exprt callee_expr;
+    if(get_expr(implicit_cast_expr, callee_expr))
+      return true;
+
+    // 2. Get type
+    // Need to "decrypt" the typeDescriptions and manually make a typeDescription
+    nlohmann::json callee_rtn_type =
+      make_callexpr_return_type(callee_expr_json["typeDescriptions"]);
+    typet type;
+    if(get_type_description(callee_rtn_type, type))
+      return true;
+
+    side_effect_expr_function_callt call;
+    call.function() = callee_expr;
+    call.type() = type;
+
+    // 3. Set side_effect_expr_function_callt
+    unsigned num_args = 0;
+    for(const auto &arg : expr["arguments"].items())
+    {
+      exprt single_arg;
+      if(get_expr(arg.value(), single_arg))
+        return true;
+
+      call.arguments().push_back(single_arg);
+      ++num_args;
+    }
+    msg.status(fmt::format("  @@ num_args={}", num_args));
+
+    // 4. Convert call arguments
+    new_expr = call;
+    break;
+  }
+  case SolidityGrammar::ExpressionT::ImplicitCastExprClass:
+  {
+    if(get_cast_expr(expr, new_expr))
+      return true;
+    break;
+  }
+  case SolidityGrammar::ExpressionT::IndexAccess:
+  {
+    // 1. get type, this is the base type of array
+    typet t;
+    if(get_type_description(expr["typeDescriptions"], t))
+      return true;
+
+    // 2. get the decl ref of the array
+    // wrap it in an ImplicitCastExpr to perform conversion of ArrayToPointerDecay
+    nlohmann::json implicit_cast_expr =
+      make_implicit_cast_expr(expr["baseExpression"], "ArrayToPointerDecay");
+    exprt array;
+    if(get_expr(implicit_cast_expr, array))
+      return true;
+
+    // 3. get the position index
+    exprt pos;
+    if(get_expr(expr["indexExpression"], pos))
+      return true;
+
+    new_expr = index_exprt(array, pos, t);
+    break;
+  }
+  default:
+  {
+    assert(!"Unimplemented type in rule expression");
+    return true;
+  }
+  }
+
+  new_expr.location() = location;
+  return false;
+}
+
+bool solidity_convertert::get_binary_operator_expr(
+  const nlohmann::json &expr,
+  exprt &new_expr)
+{
+  // preliminary step for recursive BinaryOperation
+  current_BinOp_type.push(&(expr["typeDescriptions"]));
+
+  // 1. Convert LHS and RHS
+  // For "Assignment" expression, it's called "leftHandSide" or "rightHandSide".
+  // For "BinaryOperation" expression, it's called "leftExpression" or "leftExpression"
+  exprt lhs, rhs;
+  if(expr.contains("leftHandSide"))
+  {
+    if(get_expr(expr["leftHandSide"], lhs))
+      return true;
+
+    if(get_expr(expr["rightHandSide"], rhs))
+      return true;
+  }
+  else if(expr.contains("leftExpression"))
+  {
+    if(get_expr(expr["leftExpression"], lhs))
+      return true;
+
+    if(get_expr(expr["rightExpression"], rhs))
+      return true;
+  }
```

Comment:
I think that you can simplify this with ternary operators and moving the assertion before e.g

```cpp
assert(expr.contains("leftHandSide") || expr.contains("leftExpression") );
if(get_expr(expr.contains("leftHandSide") ? expr["leftHandSide"] : expr["leftExpression"], lhs))
      return true;
```